src/Aspire.Dashboard/Mcp/ResourceMcpTools.cs

```diff
@@ -0,0 +1,384 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.ComponentModel;
+using System.Diagnostics.CodeAnalysis;
+using Aspire.Dashboard.ConsoleLogs;
+using Aspire.Dashboard.Model;
+using Aspire.Dashboard.Model.Assistant;
+using Aspire.Dashboard.Model.Otlp;
+using Aspire.Dashboard.Otlp.Model;
+using Aspire.Dashboard.Otlp.Storage;
+using Aspire.Hosting.ConsoleLogs;
+using ModelContextProtocol;
+using ModelContextProtocol.Server;
+
+namespace Aspire.Dashboard.Mcp;
+
+[McpServerToolType]
+internal sealed class ResourceMcpTools
+{
+    private static readonly ConcurrentDictionary<string, OtlpTrace> s_referencedTraces = new();
+    private static readonly ConcurrentDictionary<long, OtlpLogEntry> s_referencedLogs = new();
```

Comment:
@halter73 In MCP is there a reliable, standard way to keep state in the MCP server for an active conversation?

For example, someone starts a conversation that uses Aspire MCP to get logs. Logs returned in the response from AI are remembered, and then when a log ID is referenced in later MCP calls in the same conversation, we can use that reference again to do some extra processing.

For this to work we'd need to know some ID for the conversation, and a way to know when it is finished so objects aren't held onto longer than they need to be.

Reply:
And in the header otherwise of HttpContext